pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java

```diff
@@ -847,6 +847,15 @@ protected boolean hasPendingBatchReceive() {
         return pendingBatchReceives != null && peekNextBatchReceive() != null;
     }
 
+    protected void resetIncomingMessageSize() {
+        INCOMING_MESSAGES_SIZE_UPDATER.set(this, 0);
+    }
+
+    protected void updateIncomingMessageSize(final Message<?> message) {
+        INCOMING_MESSAGES_SIZE_UPDATER.addAndGet(this,
+                (message.getData() != null) ? message.getData().length : 0);
+    }
```

Comment:
just wondering if `message.getData().length` this should be negated as it's in the original code that it's replacing?

Reply:
@BewareMyPower Could you please take a look @lhotari 's comment?

Reply:
I think the previous implementation is wrong but there're no tests that expose the bug. The value only affects https://github.com/apache/pulsar/blob/d4c1677be93e5673672018f79da8a213514f01a3/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java#L673-L679

If it's negative, `INCOMING_MESSAGES_SIZE_UPDATER.get(this)` will be always negative and the comparison with `batchReceivePolicy.getMaxNumBytes()` is meaningless. I noticed the change was involved in #4621 , could you also take a look?

Reply:
Hi @lhotari @BewareMyPower, I will push a fix soon. We should decrease the incoming queue size when take messages from the queue and increase the incoming queue size while adding messages to the queue. With #9113, will always increase the incoming queue size.